dark/diamond/conversion.py

```diff
@@ -0,0 +1,286 @@
+from __future__ import print_function
+
+import bz2
+from json import dumps, loads
+from operator import itemgetter
+
+from Bio.File import as_handle
+
+from dark.hsp import HSP, LSP
+from dark.score import HigherIsBetterScore
+from dark.alignments import Alignment, ReadAlignments
+from dark.diamond.hsp import normalizeHSP
+
+
+class DiamondTabularFormatReader(object):
+    """
+    Provide a method that yields parsed tabular records from a file. Store and
+    make accessible the global BLAST parameters.
+    Make sure you run DIAMOND with the right output flags set. The flags are:
+    qseqid, sseqid, bitscore, evalue, qframe, qseq, qstart, qend, sseq, sstart,
+    send, slen.
+
+    @param filename: A C{str} filename or an open file pointer, containing XML
+        BLAST records.
+    """
+
+    def __init__(self, filename):
+        self._filename = filename
+        self.params = {
+            'application': 'DIAMOND',
+            'version': 'v0.8.23',
+            'reference': ('Buchfink et al., Fast and Sensitive Protein '
+                          'Alignment using DIAMOND, Nature Methods, 12, 59–60 '
+                          '(2015)')
+        }
+
+    def records(self):
+        """
+        Parse the DIAMOND output file and yield records.
+        """
+        with as_handle(self._filename) as fp:
+            previousQseqid = None
+            subjectsSeen = {}
+            record = {}
+            for line in fp:
+                (qseqid, sseqid, bitscore, evalue, qframe, qseq, qstart, qend,
+                 sseq, sstart, send, slen) = line.split('\t')
+                if previousQseqid == qseqid:
+                    # We have already started accumulating alignments for this
+                    # query.
+                    if sseqid not in subjectsSeen:
+                        # We have not seen this subject before, so this is
+                        # a new alignment.
+                        subjectsSeen.add(sseqid)
+                        hsp = {
```

Comment:
I'd move the setting of `hsp` up before the `if` on line 51. Then you don't have to duplicate this code below (on line 75). You need `hsp` in either branch of the `if``.